src/gui/programupdater.cpp

```diff
@@ -165,11 +169,8 @@ void ProgramUpdater::rssDownloadFinished(const Net::DownloadResult &result)
         }
     }
 
-    emit updateCheckFinished(!m_updateUrl.isEmpty(), version, m_invokedByUser);
-}
-
-void ProgramUpdater::updateProgram() const
-{
-    Q_ASSERT(!m_updateUrl.isEmpty());
-    QDesktopServices::openUrl(m_updateUrl);
+    if (updateUrl.isEmpty())
+        emit updateCheckFinished(false, m_invokedByUser, version, []() {});
+    else
+        emit updateCheckFinished(true, m_invokedByUser, version, [updateUrl]() { updateProgram(updateUrl); });
```

Comment:
The following came to mind when I read this:
```c++
[updateUrl{std::move(updateUrl)}]() { updateProgram(updateUrl); }
```

I will leave it up to you if you want to change it.
And google gave me this: https://stackoverflow.com/questions/8640393/move-capture-in-lambda

Reply:
Also why not:
```c++
[updateUrl{std::move(updateUrl)}]() { QDesktopServices::openUrl(updateUrl); }
```

Reply:
>I will leave it up to you if you want to change it.

I would like to avoid `std::move` unless it is necessary, just some personal preference.

>Also why not:

I suppose there can be another abstraction for the update action (whether it be open url or really updates the program executable).

Reply:
@Chocobo1, @sledgehammer999
What you are discussing here is even more incorrect than using `sender()` method inside slot. I strongly disapprove of this.
Even if "open URL" is considered as an update logic (assuming that it may become something more advanced at some point), so that it should be inside the ProgramUpdater class, it should not interact in such a confusing way. There are still two different (although related) actions, "checking for updates" and "updating". When "checking for updates" is finished ProgramUpdater we need to report the results to the user and get confirmation to perform the update. Currently ProgramUpdater emits a signal when checking is finished. But emitting signal shouldn't force observers to perform some particular action that observable wants. This is a pattern violation, etc. (it leads to incorrect design/architecture). All ProgramUpdater should do is to report on the event. Control logic must be outside of it. This is true in the existing code, except for the way ProgramUpdater is accessed inside the slot (using `sender()` method). It should be either directly accessible at this scope (i.e. it should be a class variable) or (if you don't want to declare it as class variable) you should use lambda as a slot and capture ProgramUpdater instance into it.
There is also another way. Just put the control logic inside the ProgramUpdater  class, so that it will show messages to the user, receive confirmation, and trigger an update itself.

Reply:
I pushed an update.

Reply:
>I pushed an update.

IMO, it's not so better than previous.
Do you disagree with the following statement?
>It should be either directly accessible at this scope (i.e. it should be a class variable) or (if you don't want to declare it as class variable) you should use lambda as a slot and capture ProgramUpdater instance into it.

Seems you can easily modify ProgramUpdater to be fully featured component (i.e. transfer all corresponding logic into it, timer etc.) and make it class variable of MainWindow (instead of current timer variable). You can also transfer confirmation logic into it too.

In case you don't want to make it you still can make the code logic clearer:
1. Move confirmation logic into function (e.g. `bool confirmUpdate()`)
2. Use lambda instead of `onUpdateCheckFinished()` and capture programUpdater instance into it (mostly pseudocode with some code omitted, just basic idea):
```c++
auto programUpdater = ...
connect(programUpdater, &ProgramUpdater::updateCheckFinished, this, [programUpdater](updateUrl, ...)
{
    if (confirmUpdate())
        programUpdater->updateProgram();
});
```
Note it's also good idea not to pass to ProgramUpdater any kind of data that meaningless for it, e.g. invokedByUser, you can keep it here too (by capturing into lambda instead).

I hope it's clear that I'm trying to outline two different solutions, and you can clearly see the boundary between them.

Reply:
>Seems you can easily modify ProgramUpdater to be fully featured component (i.e. transfer all corresponding logic into it, timer etc.) and make it class variable of MainWindow (instead of current timer variable). 

Moving timer into ProgramUpdater is possible but I think that would complicate the interface of ProgramUpdater as we currently wait for user response before restarting the timer, i.e. we don't want user to see 10 "update is available" dialogs opened.

>You can also transfer confirmation logic into it too.

I disagree ProgramUpdater would be a GUI element, IMO it is possible to be run without GUI although we currently don't do it. The class should only handle checking for new versions and updating/downloading new program.

>Move confirmation logic into function (e.g. bool confirmUpdate())

This is basically what `handleUpdateCheckFinished` is doing except (if I understand correctly) you proposed a synchronous way to show the dialogs and ask for response which is not recommended (because of downsides of `exec()`).

>Note it's also good idea not to pass to ProgramUpdater any kind of data that meaningless for it, e.g. invokedByUser, you can keep it here too (by capturing into lambda instead).

Now `invokedByUser` exists only in GUI code.

>I hope it's clear that I'm trying to outline two different solutions, and you can clearly see the boundary between them.

It is clear but you must think it in asynchronous way, otherwise you bring the downsides of `exec()` back.

Reply:
>Do you disagree with the following statement?
>
>>It should be either directly accessible at this scope (i.e. it should be a class variable) or (if you don't want to declare it as class variable) you should use lambda as a slot and capture ProgramUpdater instance into it.

By making `ProgramUpdater updater` a class variable, I had trouble capturing/passing `invokedByUser` into it and later I realized it would be easier to achieve my goal by going "non-class variable" way.